lib/evmmax/evmmax.cpp

```diff
@@ -8,6 +8,342 @@ using namespace intx;
+    //   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107837
+    alignas(IntT) std::byte aligned_storage[sizeof(IntT)] = {};
+    std::memcpy(&aligned_storage[sizeof(IntT) - src_size], src, src_size);
+    // TODO(C++23): Use std::start_lifetime_as<uint256>().
+    return to_big_endian(*reinterpret_cast<const IntT*>(&aligned_storage));
+}
+
+template <typename T>
+inline void store(uint8_t* dst, const T& x, size_t dst_size) noexcept
+{
+    const auto d = to_big_endian(x);
+    std::memcpy(dst, &reinterpret_cast<const std::byte*>(&d)[sizeof(T) - dst_size], dst_size);
+}
+
+inline constexpr int64_t compute_addmodx_cost(size_t value_size_mult) noexcept
+{
+    constexpr int64_t ADDMODX_GAS_A = 20;
+    constexpr int64_t ADDMODX_GAS_B = 15;
+
+    return (ADDMODX_GAS_A * static_cast<int64_t>(value_size_mult) + ADDMODX_GAS_B + 49) / 100;
+}
+
+inline constexpr int64_t compute_mulmodx_cost(size_t value_size_mult) noexcept
+{
+    constexpr int64_t MULMODX_LO_GAS_A = 9;
+    constexpr int64_t MULMODX_LO_GAS_B = 0;
+    constexpr int64_t MULMODX_LO_GAS_C = 24;
+
+    return (MULMODX_LO_GAS_A * static_cast<int64_t>(value_size_mult * value_size_mult) +
+               MULMODX_LO_GAS_B * static_cast<int64_t>(value_size_mult) + MULMODX_LO_GAS_C + 99) /
+           100;
+}
+
+template <typename UintT>
+struct EXMMAXModState : public EXMMAXModStateInterface
+{
+    std::vector<UintT> values;
+    const ModArith<UintT> arith;
+    const size_t value_size_mult;
+    const int64_t addmodx_cost;
+    const int64_t mulmodx_cost;
+
+    explicit EXMMAXModState(const UintT& mod, size_t mod_size, size_t vals_used) noexcept
+      : arith(mod),
+        value_size_mult((mod_size + 7) / 8),
+        addmodx_cost(compute_addmodx_cost(value_size_mult)),
+        mulmodx_cost(compute_mulmodx_cost(value_size_mult))
+    {
+        values.resize(vals_used);
+    }
+
+    [[nodiscard]] bool loadx(
+        uint8_t* out_ptr, size_t val_idx, size_t num_vals) const noexcept override
+    {
+        if (val_idx + num_vals > values.size())
+            return false;
+
+        for (unsigned i = 0; i < num_vals; ++i)
+            store(out_ptr + i * value_size_mult * 8, arith.from_mont(values[val_idx + i]),
+                value_size_mult * 8);
+
+        return true;
+    }
+
+    [[nodiscard]] bool storex(
+        const uint8_t* in_ptr, size_t dst_val_idx, size_t num_vals) noexcept override
+    {
+        if (dst_val_idx + num_vals > values.size())
+            return false;
+
+        for (unsigned i = 0; i < num_vals; ++i)
+        {
+            values[dst_val_idx + i] =
+                arith.to_mont(load<UintT>(in_ptr + value_size_mult * 8 * i, value_size_mult * 8));
+        }
+
+        return true;
+    }
+
+    [[nodiscard]] bool addmodx(size_t dst_idx, size_t x_idx, size_t y_idx) noexcept override
+    {
+        if (dst_idx < values.size() && x_idx < values.size() && y_idx < values.size())
```

Comment:
I think we should assume the indexes are valid (because they are going to be validated by EOF). Debug asserts are fine obviously.